ERCs/ERC5XXX draft.md

```diff
@@ -0,0 +1,173 @@
+2. The token issuer samples a `scriptPrivKey` and associated `scriptKeyAddr`. They then sign `script`, to get a signature `sigScript` and stores both at one or more URIs, which are consolidated to a scriptURI.
+3. The token issuer calls `setVerificationKey(scriptKeyAddr)` on the smart contract to set the initial scriptURI signing key.
+4. The token issuer signs the scriptURI with the signing key `scriptPrivKey`. Denote the signature `sigScriptURI`. 
+4. The token issuer calls `setScriptURI` with the scriptURI, along with `sigScriptURI`.
+
+- Update script
+1. The token issuer signs a new script `script2` using `scriptPrivKey` to get a signature `sigScript2`. 
+2. The token issuer updates `script` and `sigScript` at all its URI locations with `script2` and `sigScript2`.
+3. Replace `script` with `script2` and `sigScript` with `sigScript2`.
+
+- Update script signing key
+1. The token issuer samples a new script signing key pair denoted `scriptPrivKey2` for the private part and `scriptKeyAddr2` for the public part.
+2. The token issuer signs `script` using `scriptPrivKey2` to get `sigScript2` and then replaces `sigScript` with `sigScript2` at all its URI locations.
+3. The token issuer calls `setVerificationKey(scriptKeyAddr2)`.
+4. Replace `scriptPrivKey` with `scriptPrivKey2`, `scriptKeyAddr` with `scriptKeyAddr2` and `sigScript` with `sigScript2`.
+
+- Update script URI
+1. The token issuer moves `script` and `sigScript` from their current URI locations to all new relevant URI locations, and based on this constructs a new scriptURI structure. 
+2. The token issuer signs the new scriptURI structure with the signing key `scriptPrivKey`. Denote the signature `sigScriptURI2`.
+3. The token issuer calls `setScriptURI` with the new scriptURI structure, along with `signScriptURI2`.
+4. Replace `scriptPrivKey` with `scriptPrivKey2`.
+
+
+### Specification
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY” and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+We define a scriptURI element using the following structs:
+
+```
+struct ScriptURI {
+    ScriptURIElement[] scriptURIElements;
+}
+struct ScriptURIElement {
+    string URIOfScript;
+    string URIOfSignature;
+}
+```
+
+Based on these elements we define the smart contract interface below:
+```
+interface IERC5XXX {
+    /// @dev This event emits when the scriptURI is updated, 
+    /// so wallets implementing this interface can update a cached script
+    event ScriptUpdate(ScriptURI memory newScriptURI);
+
+    /// @dev This event emits when the script/scriptURI signing key is updated, 
+    /// so wallets implementing this interface can update a cached script
+    event VerificationKeyUpdate(address memory newVerificationKey);
+
+
+    /// @notice Get the scriptURI for the contract
+    /// @return The scriptURI
+    function getScriptURI() external view returns(ScriptURI memory);

```

Comment:
James asked me to read:

Should `getScriptURI()` be `scriptURI()` instead? More idiomatic and mentioned in other parts of this doc.

Seems like it would be better to narrow down `IERC5XXX` to just:

```
event ScriptUpdate(ScriptURI memory newScriptURI);
function getScriptURI() external view returns(ScriptURI memory);
```

or even:

```
function getScriptURI() external view returns(ScriptURI memory);
```

The rest looks optional and might slow adoption. If we can draw the line somewhere, the rest of the stuff can be in the reference implementation and submitted to OpenZeppelin. Without the parts that covers signature, there's a chance someone could use it in a non-TokenScript way, then the TokenScript engine or the client will need to innspect the contents referred to by the URI returned to see if it's a TokenScript file (or we might need an additional ERC165 check, but seems unnecessary).

Reply:
> Should `getScriptURI()` be `scriptURI()` instead? More idiomatic and mentioned in other parts of this doc.

I actually changed it to `getScriptURI()` since I thought it was more idiomatic 😅 But feel free to change it, if you think `scriptURI()` is better.

Reply:
> The rest looks optional and might slow adoption. If we can draw the line somewhere, the rest of the stuff can be in the reference implementation and submitted to OpenZeppelin. Without the parts that covers signature, there's a chance someone could use it in a non-TokenScript way, then the TokenScript engine or the client will need to innspect the contents referred to by the URI returned to see if it's a TokenScript file (or we might need an additional ERC165 check, but seems unnecessary).

It is a good point that it might be too broad! However, I do think that the issue we are trying to handle as a minimum requires a possibility to update the script. When it comes, to the script signing key, it could be sufficient to just use the smart contract signing key, but from the discussion with @weiwu-zhang I got the understanding that something smart was required for this ERC.
@weiwu-zhang what do you think? Should we make the script signing key optional?

In any case, it might be an idea to write a partial class instead of simply an interface, such that it can simply be copy-pasted into the base of any contract.

Reply:
> Should getScriptURI() be scriptURI() instead? More idiomatic and mentioned in other parts of this doc.
> I actually changed it to getScriptURI() since I thought it was more idiomatic 😅 But feel free to change it, if you think scriptURI() is better.

It's more idiomatic because of the similarity to [ERC721](https://eips.ethereum.org/EIPS/eip-721).

There is this snippet in that page:

```
interface ERC721Metadata /* is ERC721 */ {
    /// @notice A descriptive name for a collection of NFTs in this contract
    function name() external view returns (string _name);

    /// @notice An abbreviated name for NFTs in this contract
    function symbol() external view returns (string _symbol);

    /// @notice A distinct Uniform Resource Identifier (URI) for a given asset.
    /// @dev Throws if `_tokenId` is not a valid NFT. URIs are defined in RFC
    ///  3986. The URI may point to a JSON file that conforms to the "ERC721
    ///  Metadata JSON Schema".
    function tokenURI(uint256 _tokenId) external view returns (string);
}
```

`tokenURI(uint256)` (and also `name()` and `symbol()`).

Also how it specifies a getter and not a setter in the specs. The OpenZeppelin contracts do contain the setter though.

I suggest to post a draft early to https://ethereum-magicians.org and request for comments. To a certain extent, as long as they like it, it's going to get more acceptance :)

Reply:
Ok, I have changed it.